### PR TITLE
Create Differencer Transformation

### PR DESCRIPTION
**Pull request by RNKuhns**
#### Reference Issues/PRs

This implements the functionality described in #935.

#### What does this implement/fix? Explain your changes.

Adds a series-to-series transformer class that lets users specify one or more differences to iteratively apply to an input time series.

#### Does your contribution introduce a new dependency? If yes, which one?

No.

#### What should a reviewer concentrate their feedback on?

- Did I follow the API design correctly (leave base `fit` method as is)?
- Potential solution for not using `use_with_predict` hyperparameter

#### Any other comments?

Once we have this in a good place, I can update the forecasting notebook to include at least one example where the difference transformer is used (and ideally improves forecasting accuracy). 

##### For all contributions
- [X] I've added myself to the [list of contributors](https://github.com/alan-turing-institute/sktime/blob/main/.all-contributorsrc).
- [X] Optionally, I've updated sktime's [CODEOWNERS](https://github.com/alan-turing-institute/sktime/blob/main/CODEOWNERS) to receive notifications about future changes to these files.
- [X] I've added unit tests and made sure they pass locally.

##### For new estimators
- [X] I've added the estimator to the online documentation.
- [ ] I've updated the existing example notebooks or provided a new one to showcase how my estimator works.

### CODE REVIEW

**Comment by RNKuhns** [comment]
@aiwalter, @fkiraly and @mloning I've got my initial crack at a difference transformer completed. I've got my implementation working with some unit tests to verify that the inverse transformation reconstructs the input data correctly, and also can convert predictions into the undifferenced unit of the original series. 

That being said, I am hoping there is a better way for the `inverse_transform` method to work so that we don't need the nuisance `use_with_predict` parameter. 

In terms of the inverse, I can think of 4 different use-cases:

1. Apply inverse transformation to entire input time series
2. Apply inverse transformation to a slice of the input time series that includes the start of the series (e.g. `series[0:some_number]`)
3. Apply inverse transformation to some slice of the input series that doesn't include the start of the series
4. Apply inverse transformation to out-of-sample predictions

Right now, the `Differencer` should return correct results for cases 1, 2 and 4. Case 3 is difficult. To reverse a difference at lag `m` you need the `m` observations prior to the slice. Its easy enough to store the first `m` observations or last `m` observations to handle cases 1, 2 and 4. But I can't think of a way to handle case 3 other than storing the entire input series. 

Use-case 3 applies to in-sample predictions, so it would be good to handle it. What are the thoughts on the differencer storing the entire input series (instead of just the first `m` and last `m` observations currently)?

**Comment by RNKuhns** [comment]
@fkiraly I'd like to add fractional differencing to this, but want to get the initial functionality working first.

**Comment by RNKuhns** [comment]
@mloning, @fkiraly or @aiwalter any ideas on the test case I keep failing. 

The failing test looks to be related to `inverse_transform` changing the tranformers `__dict__`. 

I have tried to see if `inverse_transform` is doing this locally and haven't seen an issue. But I also might not be fully understanding what the check is doing. I'm under the assumption it is doing something like below to verify that `inverse_transform` doesn't update the `transformer.__dict__`. 

```python

from sktime.transformations.series.difference import Differencer
from sktime.datasets import load_airline
y = load_airline()
transformer = Differencer(lags=np.array([1]))
y_hat = transformer.fit_transform(y)

p = transformer.__dict__.copy()

y_reconstructed = transformer.inverse_transform(y_hat)

p2 = transformer.__dict__

p == p2
```

**Comment by fkiraly** [comment]
>@fkiraly I'd like to add fractional differencing to this, but want to get the initial functionality working first.

Very sensible, I agree with you. Let's not over-scope at the start.

**Comment by fkiraly** [comment]
>In terms of the inverse, I can think of 4 different use-cases:

I think indeed 1,2,4 are the most common and can be handled in the way you say.
When would 3 happen? Seems a bit artificial. We could simply return a warning saying that `inverse_transform` is valid only if certain conditions are met? Or an informative error message?

**Comment by RNKuhns** [comment]
@fkiraly the only way I see case 3 applying is in-sample predictions. Then you'd need to inverse transform a slice of the original series. 

To do that I can't think of a way other than storing the entire input series (instead of just the first and last observations). 

What are your thoughts on having the `Differencer` store the entire input series in a `_y` attribute?

If I do that then we need logic that is something to the effect of:
1. If the series to be inverse transform has same exact index as `_y` then return `_y` 
2. Otherwise, get the last `n` observations from `_y` whose index comes prior to  the first index of the series to be inverse transformed

That would work for all 4 cases and allow us to get rid of the `use_with_predict`.

**Review by mloning — COMMENTED** [review]
Thanks @RNKuhns, I left a few comments on the code. 

Perhaps it's easiest if we simply store the entire series and raise an error if there is no prior data available for the data passed to `inverse_transform`. The storing would have to happen in `fit`. 

Alternatively, there is the `fit-in-transform` for handling some transformers that do some fitting during `transform` but I think it's preferable to use `fit` in this case.

File: sktime/transformations/series/difference.py [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+import numpy as np
+import pandas as pd
+from sklearn.utils import check_array
+
+from sktime.transformations.base import _SeriesToSeriesTransformer
+from sktime.utils.validation import is_int
+from sktime.utils.validation.series import check_series
+
+
+def _check_differencer_lags(lags):
```
[the remark this answers is not in the document]
Comment (RNKuhns) [reply sktime/transformations/series/difference.py:17]: I've made this change.

File: sktime/transformations/series/difference.py [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+    """
+
+    _tags = {
+        "fit-in-transform": True,
+        "transform-returns-same-time-index": True,
+        "univariate-only": True,
+    }
+
+    def __init__(self, lags=1, use_with_predict=False):
+        self.lags = _check_differencer_lags(lags)
```
[the remark this answers is not in the document]
Comment (RNKuhns) [reply sktime/transformations/series/difference.py:88]: I've moved the check to fit.

File: sktime/transformations/series/difference.py [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+        self.use_with_predict = use_with_predict
+        self._cumalative_lags = self.lags.cumsum()
```
[the remark this answers is not in the document]
Comment (RNKuhns) [reply sktime/transformations/series/difference.py:90]: I've initialized the attributes as None in the constructor and I'm initializing the arrays in fit now.

**Comment by RNKuhns** [comment]
@mloning I've updated the Differencer code. Note that this is failing a test in `test_all_transformers` because `inverse_transform` doesn't match original series. It reverses the transformation, but by default the `transform` method removes the first `n` observations that are missing due to the differencing; therefore, `inverse_transform`  correctly reverses the transformation, but doesn't have the first `n` observations (as expected).

Note that the Differencer has the tag "transform-returns-same-time-index": False.  When that tag is false, the tests I've copied from `test_all_transformers.py` might not apply. 

```python
def check_transform_returns_same_time_index(Estimator):
    assert issubclass(Estimator, _SeriesToSeriesTransformer)
    estimator = _construct_instance(Estimator)
    fit_args = _make_args(estimator, "fit")
    estimator.fit(*fit_args)
    for method in ["transform", "inverse_transform"]:
        if hasattr(estimator, method):
            X = _make_args(estimator, method)[0]
            Xt = estimator.transform(X)
            np.testing.assert_array_equal(X.index, Xt.index)


def check_transform_inverse_transform_equivalent(Estimator):
    estimator = _construct_instance(Estimator)
    X = _make_args(estimator, "fit")[0]
    Xt = estimator.fit_transform(X)
    Xit = estimator.inverse_transform(Xt)
    _assert_array_almost_equal(X, Xit)

```

File: sktime/transformations/series/difference.py [hunk 4]
```diff
@@ -0,0 +1,317 @@
...
+
+        is_contained_by_fitted_z = False
+        is_future = False
+
+        if first_idx < orig_first_idx:
+            msg = [
+                "Some indices of `Z` are prior to timeseries used in `fit`.",
+                "Reconstruction via `inverse_transform` is not possible.",
+            ]
+            raise ValueError(" ".join(msg))
```
Comment (mloning) [inline sktime/transformations/series/difference.py:118]: What's the different compared to passing the `msg` directly?
```suggestion
            raise ValueError(msg)
```

File: sktime/transformations/series/difference.py [hunk 5]
```diff
@@ -0,0 +1,317 @@
...
+            is_contained_by_fitted_z = True
+
+        elif first_idx > orig_last_idx:
+            is_future = True
+
+        pad_z_inv = self.remove_missing or is_future
+
+        cutoff = Z.index[0] if pad_z_inv else Z.index[self._cumulative_lags[-1]]
+        fh = ForecastingHorizon(
+            np.array([*range(-1, -(self._cumulative_lags[-1] + 1), -1)])
```
Comment (mloning) [inline sktime/transformations/series/difference.py:130]: Would that work?
```suggestion
            np.arange(-1, -(self._cumulative_lags[-1] + 1), -1)
```

File: sktime/transformations/series/difference.py [hunk 6]
```diff
@@ -0,0 +1,317 @@
...
+        )
+
+        index = fh.to_absolute(cutoff).to_pandas()
+        index_diff = index.difference(self._Z.index)
+
+        if index_diff.shape[0] != 0:
```
Comment (mloning) [inline sktime/transformations/series/difference.py:136]: in which case would that happen?

File: sktime/transformations/series/difference.py [hunk 7]
```diff
@@ -0,0 +1,317 @@
...
+
+        lags : np.ndarray
+            Lags to be used in applying differences.
+
+        Returns
+        -------
+        diff :
+            Differenced series.
+        """
+        diff = Z.copy()
```
Comment (mloning) [inline sktime/transformations/series/difference.py:185]: for readability, so that we don't have `diff.diff(...)`
```suggestion
        Zt = Z.copy()
```

File: sktime/transformations/series/difference.py [hunk 8]
```diff
@@ -0,0 +1,317 @@
...
+
+        if len(lags) == 0:
+            return diff
+
+        else:
+            for lag in lags:
+                diff = diff.diff(lag)
```
Comment (mloning) [inline sktime/transformations/series/difference.py:192]: what's the scenario where you want to apply different lags iteratively?

**Review by mloning — COMMENTED** [review]
Thanks @RNKuhns - do you want to skip those tests based on the tags by adding an if condition in the test function you posted above? I think that would make sense.

**Review by RNKuhns — COMMENTED** [review]
> Thanks @RNKuhns - do you want to skip those tests based on the tags by adding an if condition in the test function you posted above? I think that would make sense.

Yep, that would be my proposed solution.

File: sktime/transformations/series/difference.py [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+    >>> from sktime.transformations.series.difference import Differencer
+    >>> from sktime.datasets import load_airline
+    >>> y = load_airline()
+    >>> transformer = Differencer()
+    >>> y_hat = transformer.fit_transform(y)
+    """
+
+    _tags = {
+        "fit-in-transform": True,
+        "transform-returns-same-time-index": True,
```
[the remark this answers is not in the document]
Comment (RNKuhns) [reply sktime/transformations/series/difference.py]: These have been updated now to resolve this.

File: sktime/transformations/series/difference.py [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+
+        Returns
+        -------
+        diff :
+            Differenced series.
+        """
+        diff = series.copy()
+        for i, lag in enumerate(self.lags):
+            prior_cum_lag = self._prior_cum_lags[i]
+            self._lag_fit_info.append(_difference_info(diff, lag, prior_cum_lag))
```
[the remark this answers is not in the document]
Comment (RNKuhns) [reply sktime/transformations/series/difference.py]: I had been using the fit-in-transform tag, but like we talked about I'm now storing info in fit.

File: sktime/transformations/series/difference.py (hunk 7 above)
Comment (RNKuhns) [reply sktime/transformations/series/difference.py:185]: That makes sense to me. I've updated.

File: sktime/transformations/series/difference.py (hunk 8 above)
Comment (RNKuhns) [reply sktime/transformations/series/difference.py:192]: Most use cases just need a single difference, but there are definitely cases where more than one difference (and at different lags) may be needed. 

The classic scenario is a series with trend and seasonality. In that case you would consider a first difference to help deal with the trend and a seasonal difference to deal with the seasonal dependence. 

The other case is if something is non-seasonal, but integrated with an order greater than one. That case you more than likely are going to apply iterative differences at the same lag (e.g. lags = [1, 1] for 2nd order). 

Some other implementations allow the flexibility to specify a difference at a single lag, but also specify an order parameter to account for the second case. But then you'd need to have 2 `Differencer's` in some pipelines to deal with the first case. 

Implementing it this way lets a user accomplish both cases with a single `Differencer` in their pipeline, which I think is nice.

I'll a note on these use cases in the docstring.

**Comment by RNKuhns** [comment]
@mloning I also realized there is another test not passing because of the `"transform-returns-same-time-index"` issue. I'm going to add the logic to the transformer tests that will handle them, but I'm not sure how to proceed on this one. 

The test `check_fit_idempotent` is actually not getting to the part that checks that the fit is actually `idempotent` because when it is first looping through the `NON_STATE_CHANGING_METHODS` it is passing a timeseries to inverse transform that has the same indices as the fitted data. But `inverse_transform` is assuming that a series that has been differenced (with lags specified in constructor). To perform the `inverse_transform` on a series that has been differenced you would need stored values prior to the series prior to the indices passed in `inverse_transform`. 

I think an example makes a bit easier to follow: 
If `check_fit_idempotent` is constructing a series starting from January 1, 2000. This series is passed in `fit` and then `transform`. Assuming `lags=1` was specified in the constructor (this generally holds but a specific example makes it easier to talk about), then `transform` will return the first-differenced series with indices starting on January 2, 2000. 

`inverse_transfrom` is expecting a first-differenced series will be passed and reconstruct the underlying series. `check_fit_idempotent` is passing a series with indices starting on January 1, 2000 to `inverse_transfrom` which is assumed to be a series of first-differenced values based on the prior `fit`. To correctly reconstruct that series you'd need to know the value on December 31, 1999 (because the value on January 1, 2000 is assumed to be a first difference) -- hence, the error being raised by the `inverse_transform` logic.

File: sktime/transformations/series/difference.py (hunk 6 above)
Comment (RNKuhns) [reply sktime/transformations/series/difference.py:136]: In the step above, `ForecastingHorizon` to calculate, based on the lags used in the differencing, the index values prior to the series passed to `inverse_transform` that would be needed to reconstruct the series. For example, a first-difference would need 1 prior indices, a difference at lag 12 (seasonal difference on monthly data) would need 12 prior indices, etc. 

`index_diff` is any indices that are in the index of indices needed for the reconstruction but not not present in the fitted data (`self._Z`). If those are not there then it is not possible to do the reconstruction.

**Comment by RNKuhns** [comment]
I ended up adding code related to the "transform-returns-same-time-index" tag to the following transformer tests:
check_transform_returns_same_time_index
check_transform_inverse_transform_equivalent
check_series_to_series_transform_multivariate

**Comment by RNKuhns** [comment]
Merged main into the branch.

File: sktime/transformations/series/difference.py [hunk 11]
```diff
@@ -0,0 +1,315 @@
...
+    >>> y_transform = transformer.fit_transform(y)
+    """
+
+    _tags = {
+        "fit-in-transform": False,
+        "transform-returns-same-time-index": False,
+        "univariate-only": False,
+    }
+
+    def __init__(self, lags=1, remove_missing=True):
```
Comment (mloning) [inline sktime/transformations/series/difference.py:114]: in line with common usage in pandas
```suggestion
    def __init__(self, lags=1, drop_na=True):
```

**Review by mloning — COMMENTED** [review]
Hi @RNKuhns, I just checked out the code locally and played around with it. 

I think there is a small bug in there somewhere. This one gives an error: 
```python
from sktime.transformations.series.difference import Differencer
from sktime.forecasting.all import *

y = load_airline()
transformer = Differencer(remove_missing=True)
yt = transformer.fit_transform(y)
yit = transformer.inverse_transform(yt)
np.testing.assert_array_equal(y, yit)
```
When switching `remove_missing=False` it works. Could we add this as a test? I'm also not sure why this hasn't been picked up by the `check_transform_inverse_transform_equivalent` automated unit test.

File: sktime/transformations/series/difference.py (hunk 5 above)
Comment (RNKuhns) [reply sktime/transformations/series/difference.py]: I belive it does and I've switched the code to that.

File: sktime/transformations/series/difference.py (hunk 11 above)
Comment (RNKuhns) [reply sktime/transformations/series/difference.py:114]: I'm down for that. Will push updated version shortly.

**Comment by RNKuhns** [comment]
@mloning on the unit testing that is actually the expected behavior. One of the tweaks in the PR was to the unit testing for transformers. The test `check_transform_inverse_transform_equivalent` needs to apply differently to transformers where the tag "transform-returns-same-time-index" is False. In that case we can only check that the indices that are in the original series and the inverse transformed series are equal.

**Comment by RNKuhns** [comment]
@mloning I know you'll be busy with the docsprint, but this is now ready for a final look when you have a chance.

**Review by mloning — APPROVED** [review]
Thanks @RNKuhns - had another look, looks all good to me, so we can merge this now!

**Comment by mloning** [comment]
Now finally merged @RNKuhns - sorry for the delay on this! 🎉
